File: frontend/src/api/metrics.ts
import { api } from "./client";
import type { OptimizationSuggestion } from "../types";

export interface TrendPoint {
  date: string;
  play: number;
  exposure: number;
}

export interface EngagementPoint {
  date: string;
  completion_rate: number;
  like_rate: number;
}

export interface RankItem {
  title: string;
  completion_rate: number;
}

export interface ReviewOverview {
  has_data: boolean;
  trend: TrendPoint[];
  engagement: EngagementPoint[];
  rank_top: RankItem[];
  rank_bottom: RankItem[];
  total_play: number;
  avg_completion_rate: number;
  follower_delta: number;
}

export interface PerformanceSnapshot {
  id: number;
  content_item_id: number | null;
  account_id: number | null;
  source: "douyin" | "xiaohongshu" | "shipinhao" | "manual" | "demo";
  stat_date: string;
  title: string | null;
  play: number;
  exposure: number;
  completion_rate: number;
  like_rate: number;
  comment_rate: number;
  share_rate: number;
  follower_delta: number;
  completion_rate_5s: number | null;
  bounce_rate_2s: number | null;
  profile_visit_count: number | null;
  created_at: string;
}

export type ReviewPeriodDays = 7 | 30 | 90;

export interface ReviewGoalInput {
  period_days: ReviewPeriodDays;
  target_play?: number;
  target_completion_rate?: number;
  target_follower_delta?: number;
}

export interface ReviewGoal {
  id: number | null;
  period_days: number;
  target_play?: number | null;
  target_completion_rate?: number | null;
  target_follower_delta?: number | null;
  status: "not_configured" | "insufficient_data" | "behind" | "on_track" | "achieved";
  achievement_percent: number | null;
  components: Array<{
    metric: "play" | "completion_rate" | "follower_delta";
    label: string;
    current: number;
    target: number;
    achievement_percent: number;
  }>;
  summary: string;
}

export interface ReviewWorkspace {
  account: {
    id: number;
    nickname: string;
    platform: string;
    auth_status: string;
    data_sync_status: string;
  };
  period: {
    days: number;
    current_start: string;
    current_end: string;
    previous_start: string;
    previous_end: string;
  };
  data_status: {
    has_data: boolean;
    sources: Array<PerformanceSnapshot["source"] | "platform_export" | "screenshot_verified" | "manual_entry" | "official_api">;
    latest_stat_date: string | null;
    latest_synced_at: string | null;
    latest_confirmed_at: string | null;
    days_since_observed: number | null;
    days_since_confirmed: number | null;
    coverage: Record<string, "available" | "missing" | "ambiguous">;
    conflict_count: number;
    source_summary: Array<{
      batch_id: number | null;
      source_kind: string;
      data_domains: string[];
      confirmed_at: string | null;
      period_start: string | null;
      period_end: string | null;
    }>;
    missing_reasons: string[];
  };
  goal: ReviewGoal;
  conclusion: string;
  totals: {
    play: number;
    exposure: number;
    avg_completion_rate: number;
    avg_engagement_rate: number;
    follower_delta: number;
  };
  changes: Array<{
    metric: "play" | "completion_rate" | "follower_delta";
    label: string;
    current: number;
    previous: number | null;
    delta_percent: number | null;
    direction: "up" | "down" | "flat" | "baseline";
    summary: string;
  }>;
  trend: TrendPoint[];
  engagement: EngagementPoint[];
  attributions: Array<{
    content_item_id: number | null;
    title: string;
    play: number;
    completion_rate: number;
    engagement_rate: number;
    role: "driver" | "opportunity";
    reason: string;
  }>;
  evidence: PerformanceSnapshot[];
  suggestions: OptimizationSuggestion[];
}

export async function getReviewOverview(days = 30): Promise<ReviewOverview> {
  const { data } = await api.get<ReviewOverview>("/metrics/overview", { params: { days } });
  return data;
}

export async function listPerformanceSnapshots(
  accountId?: number | null,
): Promise<PerformanceSnapshot[]> {
  const { data } = await api.get<PerformanceSnapshot[]>("/metrics/performance-snapshots", {
    params: accountId != null ? { account_id: accountId } : undefined,
  });
  return data;
}

export async function getReviewWorkspace(
  accountId: number,
  days: ReviewPeriodDays,
): Promise<ReviewWorkspace> {
  const { data } = await api.get<ReviewWorkspace>("/metrics/review-workspace", {
    params: { account_id: accountId, days },
  });
  return data;
}

export async function upsertReviewGoal(
  accountId: number,
  input: ReviewGoalInput,
): Promise<ReviewGoal> {
  const { data } = await api.put<ReviewGoal>(`/metrics/review-goals/${accountId}`, input);
  return data;
}

File: backend/app/schemas/metrics.py
"""复盘看板 schema：指标录入 + 聚合视图。"""

from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.models.enums import MetricSource
from app.schemas.feedback import OptimizationSuggestionOut


class IngestMetricRequest(BaseModel):
    """录入/回流一条指标快照（E8 抖音回流与手动录入共用）。"""

    content_item_id: int | None = None
    account_id: int | None = None
    source: MetricSource = MetricSource.MANUAL
    stat_date: date
    title: str | None = Field(default=None, max_length=300)
    play: int = 0
    exposure: int = 0
    completion_rate: float = 0.0
    like_rate: float = 0.0
    comment_rate: float = 0.0
    share_rate: float = 0.0
    follower_delta: int = 0


class TrendPoint(BaseModel):
    date: str
    play: int
    exposure: int


class EngagementPoint(BaseModel):
    date: str
    completion_rate: float
    like_rate: float


class RankItem(BaseModel):
    title: str
    completion_rate: float


class ReviewOverview(BaseModel):
    """复盘聚合视图：趋势 + 完播互动 + 内容排名 + 汇总。"""

    model_config = ConfigDict(from_attributes=True)

    has_data: bool
    trend: list[TrendPoint]
    engagement: list[EngagementPoint]
    rank_top: list[RankItem]
    rank_bottom: list[RankItem]
    total_play: int
    avg_completion_rate: float
    follower_delta: int


class PerformanceSnapshotOut(BaseModel):
    """A normalized content performance snapshot for review loops."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    content_item_id: int | None
    account_id: int | None
    source: MetricSource
    stat_date: date
    title: str | None
    play: int
    exposure: int
    completion_rate: float
    like_rate: float
    comment_rate: float
    share_rate: float
    follower_delta: int
    created_at: datetime


class ReviewGoalUpsert(BaseModel):
    period_days: Literal[7, 30, 90]
    target_play: int | None = Field(default=None, ge=1)
    target_completion_rate: float | None = Field(default=None, gt=0, le=1)
    target_follower_delta: int | None = Field(default=None, ge=1)

    @model_validator(mode="after")
    def require_target(self):
        if not any(
            value is not None
            for value in (
                self.target_play,
                self.target_completion_rate,
                self.target_follower_delta,
            )
        ):
            raise ValueError("至少设置一项周期目标")
        return self


class AccountReviewGoalOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int | None = None
    period_days: int
    target_play: int | None = None
    target_completion_rate: float | None = None
    target_follower_delta: int | None = None
    status: Literal["not_configured", "insufficient_data", "behind", "on_track", "achieved"]
    achievement_percent: float | None = None
    components: list[dict] = Field(default_factory=list)
    summary: str


class ReviewAccountOut(BaseModel):
    id: int
    nickname: str
    platform: str
    auth_status: str
    data_sync_status: str


class ReviewPeriodOut(BaseModel):
    days: int
    current_start: date
    current_end: date
    previous_start: date
    previous_end: date


class ReviewDataStatusOut(BaseModel):
    has_data: bool
    sources: list[MetricSource]
    latest_stat_date: date | None = None
    latest_synced_at: datetime | None = None
    missing_reasons: list[str] = Field(default_factory=list)


class ReviewTotalsOut(BaseModel):
    play: int = 0
    exposure: int = 0
    avg_completion_rate: float = 0
    avg_engagement_rate: float = 0
    follower_delta: int = 0


class ReviewChangeOut(BaseModel):
    metric: Literal["play", "completion_rate", "follower_delta"]
    label: str
    current: float
    previous: float | None = None
    delta_percent: float | None = None
    direction: Literal["up", "down", "flat", "baseline"]
    summary: str


class ReviewAttributionOut(BaseModel):
    content_item_id: int | None = None
    title: str
    play: int
    completion_rate: float
    engagement_rate: float
    role: Literal["driver", "opportunity"]
    reason: str


class ReviewWorkspaceOut(BaseModel):
    account: ReviewAccountOut
    period: ReviewPeriodOut
    data_status: ReviewDataStatusOut
    goal: AccountReviewGoalOut
    conclusion: str
    totals: ReviewTotalsOut
    changes: list[ReviewChangeOut]
    trend: list[TrendPoint]
    engagement: list[EngagementPoint]
    attributions: list[ReviewAttributionOut]
    evidence: list[PerformanceSnapshotOut]
    suggestions: list[OptimizationSuggestionOut]
